File: src/main/python/AoC2019_04.py
#! /usr/bin/env python3
#
# Advent of Code 2019 Day 3
#

import sys
from collections import Counter
from collections.abc import Callable

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.range import RangeInclusive

Input = RangeInclusive
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        a, b = next(iter(input_data)).split("-")
        return RangeInclusive.between(int(a), int(b))

    def does_not_decrease(self, passw: str) -> bool:
        return list(passw) == sorted(passw)

    def is_valid_1(self, passw: str) -> bool:
        return self.does_not_decrease(passw) and len(set(passw)) != len(passw)

    def is_valid_2(self, passw: str) -> bool:
        return self.does_not_decrease(passw) and 2 in Counter(passw).values()

    def count_valid(self, rng: Input, check: Callable[[str], bool]) -> int:
        return sum(check(str(i)) for i in rng.iterator())

    def part_1(self, rng: Input) -> int:
        return self.count_valid(rng, self.is_valid_1)

    def part_2(self, rng: Input) -> int:
        return self.count_valid(rng, self.is_valid_2)

    def samples(self) -> None:
        assert self.is_valid_1("122345")
        assert self.is_valid_1("111123")
        assert self.is_valid_1("111111")
        assert not self.is_valid_1("223450")
        assert not self.is_valid_1("123789")
        assert self.is_valid_2("112233")
        assert not self.is_valid_2("123444")
        assert self.is_valid_2("111122")


solution = Solution(2019, 4)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2019_04.java
import static com.github.pareronia.aoc.Utils.asCharacterStream;
import static java.util.stream.Collectors.toSet;

import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;

import com.github.pareronia.aoc.Counter;
import com.github.pareronia.aoc.IntegerSequence.Range;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2019_04 extends SolutionBase<Range, Long, Long> {
    
    private AoC2019_04(final boolean debug) {
        super(debug);
    }

    public static AoC2019_04 create() {
        return new AoC2019_04(false);
    }

    public static AoC2019_04 createDebug() {
        return new AoC2019_04(true);
    }

    @Override
    protected Range parseInput(final List<String> inputs) {
        final String[] split = inputs.get(0).split("-");
        return Range.between(Integer.parseInt(split[0]), Integer.parseInt(split[1]));
    }

    private boolean doesNotDecrease(final String passw) {
        final char[] sorted = passw.toCharArray();
        Arrays.sort(sorted);
        for (int i = 0; i < passw.length(); i++) {
            if (passw.charAt(i) != sorted[i]) {
                return false;
            }
        }
        return true;
    }
    
    private boolean isValid1(final String string) {
        return doesNotDecrease(string) &&
            asCharacterStream(string).collect(toSet()).size() != string.length();
    }
    
    private boolean isValid2(final String string) {
        return doesNotDecrease(string) &&
            new Counter<>(asCharacterStream(string)).containsValue(2L);
    }
    
    private long countValid(final Range range, final Predicate<String> isValid) {
        return range.intStream()
                .mapToObj(String::valueOf)
                .filter(isValid::test)
                .count();
    }

    @Override
    public Long solvePart1(final Range range) {
        return countValid(range, this::isValid1);
    }

    @Override
    public Long solvePart2(final Range range) {
        return countValid(range, this::isValid2);
    }

    @Override
    public void samples() {
        final AoC2019_04 test = AoC2019_04.createDebug();
        assert test.isValid1("122345");
        assert test.isValid1("111123");
        assert test.isValid1("111111");
        assert !test.isValid1("223450");
        assert !test.isValid1("123789");
        assert test.isValid2("112233");
        assert !test.isValid2("123444");
        assert test.isValid2("111122");
    }

    public static void main(final String[] args) throws Exception {
        AoC2019_04.create().run();
    }
}